File: codes/perl/Array.map.pl
use strict;
use warnings;
use JSON;
use Scalar::Util qw(looks_like_number);

sub pretty_array_of_primitives {
    my $number_of_parameters = @_;
    my $result = "[";
    for (my $array_item_index = 0; $array_item_index < $number_of_parameters; $array_item_index += 1) {
        my $array_item = $_[$array_item_index];
        my $is_string = defined($array_item) && $array_item =~ /[0-9a-zA-Z`~!@#%&_=;':", \(\)\[\]\{\}\|\*\+\?\^\$\/\\\<\>\.\-]/;
        my $is_number = looks_like_number($array_item);
        last if (!$is_string && !$is_number);
        $result = $result . "\"" . $array_item . "\"" if ($is_string);
        $result = $result . $array_item if ($is_number);
        $result = $result . ", " if (($array_item_index + 1) != $number_of_parameters);
    }
    $result = $result . "]";
    return $result;
}

sub spread_syntax_object_v1 {
    my %new_object;
    my $number_of_parameters = @_;
    for (my $i = 0; $i < $number_of_parameters; $i += 2) {
        my $object_key = $_[$i];
        my $object_value = $_[$i + 1];
        $new_object{$object_key} = $object_value;
    }
    return %new_object;
}

sub array_map_v1 {
    # JavaScript-like Array.map() function
    my ($callback_function, $an_array_ref) = @_;
    my @new_array = ();
    my @an_array = @$an_array_ref;
    for my $array_item_index (0..$#an_array) {
        my $array_item = $an_array[$array_item_index];
        my $new_array_item = $callback_function->($array_item, $array_item_index, \@an_array);
        if (ref $new_array_item eq "ARRAY") {
            my @new_array_item_as_an_array = @$new_array_item;
            my $new_array_item_as_an_array_array_length = scalar(@new_array_item_as_an_array);
            next if ($new_array_item_as_an_array_array_length == 0);
            push(@new_array, $new_array_item);
            next;
        }
        push(@new_array, $new_array_item);
    }
    return @new_array;
}

print("\n# JavaScript-like Array.map() in Perl Array\n");

my @numbers = (12, 34, 27, 23, 65, 93, 36, 87, 4, 254);
print("numbers: ", pretty_array_of_primitives(@numbers), "\n");

my @numbers_labeled;

print("# using JavaScript-like Array.map() function \"array_map_v1\"\n");

@numbers_labeled = array_map_v1(sub { my ($number) = @_; return { "$number" => ((($number % 2) == 0) ? "even" : "odd") } }, \@numbers);
print("labeled numbers ", JSON->new->allow_nonref->pretty->encode(\@numbers_labeled));
# labeled numbers: [
#     {
#         "12": "even"
#     },
#     {
#         "34": "even"
#     },
#     {
#         "27": "odd"
#     },
#     {
#         "23": "odd"
#     },
#     {
#         "65": "odd"
#     },
#     {
#         "93": "odd"
#     },
#     {
#         "36": "even"
#     },
#     {
#         "87": "odd"
#     },
#     {
#         "4": "even"
#     },
#     {
#         "254": "even"
#     }
# ]

print("# using Perl Array.map() built-in function \"map\"\n");

@numbers_labeled = map { my $number = $_; { "$number" => ((($number % 2) == 0) ? "even" : "odd") } } @numbers;
print("labeled numbers ", JSON->new->allow_nonref->pretty->encode(\@numbers_labeled));
# labeled numbers: [
#     {
#         "12": "even"
#     },
#     {
#         "34": "even"
#     },
#     {
#         "27": "odd"
#     },
#     {
#         "23": "odd"
#     },
#     {
#         "65": "odd"
#     },
#     {
#         "93": "odd"
#     },
#     {
#         "36": "even"
#     },
#     {
#         "87": "odd"
#     },
#     {
#         "4": "even"
#     },
#     {
#         "254": "even"
#     }
# ]

print("\n# JavaScript-like Array.map() in Perl Array of Hashes\n");

my @products = (
    {
        "code" => "pasta",
        "price" => 321
    },
    {
        "code" => "bubble_gum",
        "price" => 233
    },
    {
        "code" => "potato_chips",
        "price" => 5
    },
    {
        "code" => "towel",
        "price" => 499
    }
);

print("products: ", JSON->new->allow_nonref->pretty->encode(\@products));

my @products_labeled;

print("# using JavaScript-like Array.map() function \"array_map_v1\"\n");

@products_labeled = array_map_v1(sub {
    my ($product) = @_;
    my %new_product = spread_syntax_object_v1(%$product, ("label" => (($product->{"price"} > 100) ? "expensive" : "cheap")));
    return \%new_product;
}, \@products);
print("labeled products: ", JSON->new->allow_nonref->pretty->encode(\@products_labeled));
# labeled products: [
#     {
#         "code": "pasta",
#         "price": 321,
#         "label": "expensive"
#     },
#     {
#         "code": "bubble_gum",
#         "price": 233,
#         "label": "expensive"
#     },
#     {
#         "code": "potato_chips",
#         "price": 5,
#         "label": "cheap"
#     },
#     {
#         "code": "towel",
#         "price": 499,
#         "label": "expensive"
#     }
# ]

print("# using Perl Array.map() built-in function \"map\"\n");

@products_labeled = map { my $product = $_; +{ spread_syntax_object_v1(%$product, ("label" => (($product->{"price"} > 100) ? "expensive" : "cheap"))) } } @products;
print("labeled products ", JSON->new->allow_nonref->pretty->encode(\@products_labeled));
# labeled products: [
#     {
#         "code": "pasta",
#         "price": 321,
#         "label": "expensive"
#     },
#     {
#         "code": "bubble_gum",
#         "price": 233,
#         "label": "expensive"
#     },
#     {
#         "code": "potato_chips",
#         "price": 5,
#         "label": "cheap"
#     },
#     {
#         "code": "towel",
#         "price": 499,
#         "label": "expensive"
#     }
# ]

File: codes/php/array_map.php
<?php

function json_stringify($anything, $optional_argument = ["pretty" => false]) {
    return (((@$optional_argument["pretty"] ?? false) === true) ? (str_replace("/\n$/", "", json_encode($anything, JSON_PRETTY_PRINT))) : (str_replace("{", "{ ", str_replace("}", " }", str_replace(",", ", ", str_replace(":", ": ", json_encode($anything)))))));
};

function array_every($callback_function, $an_array) {
    // JavaScript-like Array.every() function array_every_v4
    foreach ($an_array as $array_item_index => $array_item) {
        if ($callback_function($array_item, $array_item_index, $an_array) === false) return false;
    }
    return true;
};

function is_like_js_null($anything) {
    return (((gettype($anything) === "null") || (gettype($anything) === "NULL")) && (is_null($anything) === true) && (isset($anything) === false));
};

function is_like_js_boolean($anything) {
    return (((gettype($anything) === "boolean") || (gettype($anything) === "bool")) && (is_bool($anything) === true) && (($anything === true) || ($anything === false)));
};

function is_like_js_string($anything) {
    return ((gettype($anything) === "string") && (is_string($anything) === true));
};

function is_like_js_numeric($anything) {
    return (((gettype($anything) === "integer") || (gettype($anything) === "int") || (gettype($anything) === "float") || (gettype($anything) === "double")) && ((is_int($anything) === true) || (is_integer($anything) === true) || (is_float($anything) === true) || (is_double($anything) === true)) && (is_numeric($anything) === true));
};

function is_like_js_array($anything) {
    return ((gettype($anything) === "array") && (is_array($anything) === true));
};

function is_like_js_object($anything) {
    if (is_like_js_array($anything) === false) return false;
    return array_every(fn($array_item) => (is_like_js_string($array_item) === true), array_keys($anything));
};

function get_type($anything) {
    if (is_like_js_null($anything) === true) return "Null";
    if (is_like_js_boolean($anything) === true) return "Boolean";
    if (is_like_js_string($anything) === true) return "String";
    if (is_like_js_numeric($anything) === true) return "Numeric";
    if (is_like_js_object($anything) === true) return "Object";
    if (is_like_js_array($anything) === true) return "Array";
    return ucwords(strtolower(gettype($anything)));
};

function string_interpolation(...$rest_arguments) {
    return array_reduce($rest_arguments,  (fn($current_result, $current_argument) => ($current_result . ((get_type($current_argument) === "String") ? ($current_argument) : (((get_type($current_argument) === "Array") && (count($current_argument) === 1)) ? (json_stringify(@$current_argument[0])) : (json_stringify($current_argument)))))), "");
};

function console_log(...$rest_arguments) {
    echo string_interpolation(...$rest_arguments) . "\n";
};

function array_map_v1($callback_function, $an_array) {
    // JavaScript-like Array.map() function array_map_v1
    $new_array = [];
    foreach ($an_array as $array_item_index => $array_item) {
        $new_array_item = $callback_function($array_item, $array_item_index, $an_array);
        $new_array = [...$new_array, $new_array_item];
    }
    return $new_array;
};

function array_map_v2($callback_function, $an_array) {
    // JavaScript-like Array.map() function array_map_v2
    $new_array = [];
    foreach ($an_array as $array_item_index => $array_item) {
        $new_array = [...$new_array, $callback_function($array_item, $array_item_index, $an_array)];
    }
    return $new_array;
};

console_log("\n// JavaScript-like Array.map() in PHP array");

$numbers = [12, 34, 27, 23, 65, 93, 36, 87, 4, 254];
console_log(string_interpolation("numbers: ", [$numbers]));

console_log('// using JavaScript-like Array.map() function "array_map_v1"');

$numbers_labeled = array_map_v1(fn($number) => [$number => ((($number % 2) === 0) ? "even" : "odd")], $numbers);
console_log(string_interpolation("labeled numbers: ", json_stringify($numbers_labeled, ["pretty" => true])));
// labeled numbers: [
//     {
//         "12": "even"
//     },
//     {
//         "34": "even"
//     },
//     {
//         "27": "odd"
//     },
//     {
//         "23": "odd"
//     },
//     {
//         "65": "odd"
//     },
//     {
//         "93": "odd"
//     },
//     {
//         "36": "even"
//     },
//     {
//         "87": "odd"
//     },
//     {
//         "4": "even"
//     },
//     {
//         "254": "even"
//     }
// ]

console_log('// using JavaScript-like Array.map() function "array_map_v2"');

$numbers_labeled = array_map_v2(fn($number) => [$number => ((($number % 2) === 0) ? "even" : "odd")], $numbers);
console_log(string_interpolation("labeled numbers: ", json_stringify($numbers_labeled, ["pretty" => true])));
// labeled numbers: [
//     {
//         "12": "even"
//     },
//     {
//         "34": "even"
//     },
//     {
//         "27": "odd"
//     },
//     {
//         "23": "odd"
//     },
//     {
//         "65": "odd"
//     },
//     {
//         "93": "odd"
//     },
//     {
//         "36": "even"
//     },
//     {
//         "87": "odd"
//     },
//     {
//         "4": "even"
//     },
//     {
//         "254": "even"
//     }
// ]

console_log('// using PHP Array.map() built-in function "array_map"');

$numbers_labeled = array_map(fn($number) => [$number => ((($number % 2) === 0) ? "even" : "odd")], $numbers);
console_log(string_interpolation("labeled numbers: ", json_stringify($numbers_labeled, ["pretty" => true])));
// labeled numbers: [
//     {
//         "12": "even"
//     },
//     {
//         "34": "even"
//     },
//     {
//         "27": "odd"
//     },
//     {
//         "23": "odd"
//     },
//     {
//         "65": "odd"
//     },
//     {
//         "93": "odd"
//     },
//     {
//         "36": "even"
//     },
//     {
//         "87": "odd"
//     },
//     {
//         "4": "even"
//     },
//     {
//         "254": "even"
//     }
// ]

console_log("\n// JavaScript-like Array.map() in PHP array of associative-arrays");

$products = [
    [
        "code" => "pasta",
        "price" => 321
    ],
    [
        "code" => "bubble_gum",
        "price" => 233
    ],
    [
        "code" => "potato_chips",
        "price" => 5
    ],
    [
        "code" => "towel",
        "price" => 499
    ]
];
console_log(string_interpolation("products: ", json_stringify($products, ["pretty" => true])));

console_log('// using JavaScript-like Array.map() function "array_map_v1"');

$products_labeled = array_map_v1(fn($product) => [...$product, "label" => ((@$product["price"] > 100) ? "expensive" : "cheap")], $products);
console_log(string_interpolation("labeled products: ", json_stringify($products_labeled, ["pretty" => true])));
// labeled products: [
//     {
//         "code": "pasta",
//         "price": 321,
//         "label": "expensive"
//     },
//     {
//         "code": "bubble_gum",
//         "price": 233,
//         "label": "expensive"
//     },
//     {
//         "code": "potato_chips",
//         "price": 5,
//         "label": "cheap"
//     },
//     {
//         "code": "towel",
//         "price": 499,
//         "label": "expensive"
//     }
// ]

console_log('// using JavaScript-like Array.map() function "array_map_v2"');

$products_labeled = array_map_v2(fn($product) => [...$product, "label" => ((@$product["price"] > 100) ? "expensive" : "cheap")], $products);
console_log(string_interpolation("labeled products: ", json_stringify($products_labeled, ["pretty" => true])));
// labeled products: [
//     {
//         "code": "pasta",
//         "price": 321,
//         "label": "expensive"
//     },
//     {
//         "code": "bubble_gum",
//         "price": 233,
//         "label": "expensive"
//     },
//     {
//         "code": "potato_chips",
//         "price": 5,
//         "label": "cheap"
//     },
//     {
//         "code": "towel",
//         "price": 499,
//         "label": "expensive"
//     }
// ]

console_log('// using PHP Array.map() built-in function "array_map"');

$products_labeled = array_map(fn($product) => [...$product, "label" => ((@$product["price"] > 100) ? "expensive" : "cheap")], $products);
console_log(string_interpolation("labeled products: ", json_stringify($products_labeled, ["pretty" => true])));
// labeled products: [
//     {
//         "code": "pasta",
//         "price": 321,
//         "label": "expensive"
//     },
//     {
//         "code": "bubble_gum",
//         "price": 233,
//         "label": "expensive"
//     },
//     {
//         "code": "potato_chips",
//         "price": 5,
//         "label": "cheap"
//     },
//     {
//         "code": "towel",
//         "price": 499,
//         "label": "expensive"
//     }
// ]
